devito/ir/support/basic.py

```diff
@@ -413,6 +413,9 @@ def distance(self, other):
                 # E.g., `self=R<f,[x, y]>`, `sai=time`,
                 #       `self.itintervals=(time, x, y)`, `n=0`
                 continue
+            elif not sai or not oai:
```

Comment:
Did you try putting it four lines above, right after its "sibling" `if`, i.e.

```
elif not sai and not oai:
    <...>
elif not sai or not ai:
    <...>
```

it'd be a bit neater...

But most importantly, what `self[n]` and `other[n]` in the failing example?

I think the condition as is is a bit flaky, because you could have `a[time ...]` and `a[3 ...]` and there would be a hazardous dependency due to a[3], but you would still enter this `if`

I'll try something and let you know